### PR TITLE
fix: widgets loading state when calculating client-side

### PR DESCRIPTION
**Pull request by AdriSolid**
#### Background
Widget loaders were not working in client-side-mode. Widget loaders should be displayed every time viewport changes, ending with the end of the computation.

#### Change List
- Add a new store state with the widget ids.
-  Add `useLoadingStateFromStore` hook and use it in each `/widgets` component.

### CODE REVIEW

File: src/redux/cartoSlice.js [hunk 1]
```diff
@@ -276,3 +294,20 @@ export const removeViewportFeatures = (data) => ({
   type: 'carto/removeViewportFeatures',
   payload: data
 });
+
+/**
+ * Action to set the widget loader state
+ * @param {object} widgetId - the id of the widget
```
Comment (VictorVelarde) [inline src/redux/cartoSlice.js:300]: types should be string & boolean right?

File: src/widgets/CategoryWidget.js [hunk 2]
```diff
@@ -25,47 +26,44 @@ function CategoryWidget(props) {
...
 
   useEffect(() => {
     const abortController = new AbortController();
-    if (
-      data &&
-      credentials &&
-      (!props.viewportFilter ||
-        (props.viewportFilter && viewportFeatures[props.dataSource]))
-    ) {
+    if (data && credentials && widgetIdIsSet) {
```
Comment (VictorVelarde) [inline src/widgets/CategoryWidget.js:41]: isLoadingStateSet ? (using widgetIdIsSet feels still not clear to me)

File: src/widgets/CategoryWidget.js [hunk 3]
```diff
@@ -94,12 +92,16 @@ function CategoryWidget(props) {
...
+      title={props.title}
+      loading={widgetLoaders[props.id]}
+      {...props.wrapperProps}
+    >
       <CategoryWidgetUI
         data={categoryData}
         formatter={props.formatter}
         labels={props.labels}
-        loading={loading}
+        loading={widgetLoaders[props.id]}
```
Comment (VictorVelarde) [inline src/widgets/CategoryWidget.js:104]: I would use `isLoading` for the prop name

Comment (AdriSolid) [reply src/widgets/CategoryWidget.js:104]: Indeed, seems better

File: src/redux/cartoSlice.js [hunk 4]
```diff
@@ -130,6 +131,23 @@ export const createCartoSlice = (initialState) => {
         if (state.viewportFeatures[sourceId]) {
           delete state.viewportFeatures[sourceId];
         }
+      },
+      setWidgetLoaders: (state, action) => {
+        const { widgetId, isLoading, displayAllLoaders } = action.payload;
+
+        if (displayAllLoaders) {
+          state.widgetLoaders = Object.fromEntries(
+            Object.keys(state.widgetLoaders).map((key) => [key, true])
```
Comment (VictorVelarde) [inline src/redux/cartoSlice.js:140]: key --> widgetId, right?

Comment (AdriSolid) [reply src/redux/cartoSlice.js:140]: Indeed! Are we overusing `widgetId` here?

File: src/widgets/useLoadingStateFromStore.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  useEffect(() => {
+    if (!widgetId) {
+      throw new Error('Widget "id" property is required.');
+    }
+
+    setLoading(filterIsByViewport);
+
+    return () => {
+      removeWidgetLoaders();
```
Comment (VictorVelarde) [inline src/widgets/useLoadingStateFromStore.js:32]: After reading this again, I guess we should remove just loading state for the current widget. If we want to keep the simmetry with the 'add' option, which allows 'all' and also one specific widgetId, we should probably allow passing a widgetId list to removeWidgetLoaders

Comment (AdriSolid) [reply src/widgets/useLoadingStateFromStore.js:32]: Let's modify our action to allow removing specific `widgetId`s, it's true that a single widget component could be un-mounted.

**Review by VictorVelarde — APPROVED** [review]
👍🏻

**Comment by VictorVelarde** [comment]
@AdriSolid I made some refactors here: 
- 'loaders' to 'loading' (everywhere)
- actions modification: `setWidgetLoadingState` / `removeWidgetLoadingState` (notice the singular) + a new `setAllWidgetsLoadingState`

Please have a look at the changes, test them to see if I broke anything and, if you are ok with it, go ahead and merge it

**Comment by AdriSolid** [comment]
@VictorVelarde nice! Will need to check in a near-future if can pass to `removeWidgetLoadingState` a `widgetId`s list.

**Pull request merged**